### PR TITLE
Add e4k connector sas utilities to v2

### PR DESCRIPTION
**Pull request by patilsnr**

### CODE REVIEW

File: iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.Authentication
+{
+    /// <summary>
+    /// Authentication method that generates shared access signature (SAS) token with refresh, based on a provided shared access key (SAK).
+    /// Build for On Behalf of scenarios only.
```
Comment (abhipsaMisra) [inline iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs:11]: We should clarify in the comments that this is for gateway scenarios when one client authenticates on behalf of another.

File: iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace Microsoft.Azure.Devices.Client.Authentication
+{
+    /// <summary>
+    /// Authentication method that generates shared access signature (SAS) token with refresh, based on a provided shared access key (SAK).
+    /// Build for On Behalf of scenarios only. 
+    /// </summary>
+    public class ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf : ClientAuthenticationWithSharedAccessKeyRefresh
+    {
+#nullable enable
```
Comment (abhipsaMisra) [inline iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs:15]: Why is this required?

File: iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        /// <summary>
+        /// Creates an instance of this class.
+        /// </summary>
+        /// <param name="sharedAccessKey">Shared access key value.</param>
+        /// <param name="parentDeviceId">Identifier of the parent device.</param>
+        /// <param name="deviceId">Device identifier.</param>
```
Comment (abhipsaMisra) [inline iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs:21]: I'm not very familiar with this scenario so I don't have the exact wordings to recommend, but it feels like we could use some additional description to clarify how the parent device relates to "this" device/module.

File: iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+        /// <summary>
+        /// Gets the shared access key name.
+        /// </summary>
+        public string ParentDeviceId { get; private set; }
+
+        ///<inheritdoc/>
+        protected override Task<string> SafeCreateNewTokenAsync(string iotHub, TimeSpan suggestedTimeToLive)
+        {
+            string sasToken = SharedAccessSignatureBuilder.GetDeviceToken(iotHub, ParentDeviceId, "$edgeHub", SharedAccessKey);
```
Comment (abhipsaMisra) [inline iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs:56]: Is there a valid use-case to make setting "$edgeHub" configurable? If not, then we should rename the class to call out that this is for edgehub scenarios only.

File: iothub/device/src/Authentication/Security/SharedAccessSignatureBuilder.cs [hunk 5]
```diff
@@ -16,6 +16,7 @@ namespace Microsoft.Azure.Devices.Client
     /// </summary>
     internal sealed class SharedAccessSignatureBuilder
     {
+        private const int DEFAULT_TOKEN_EXPIRY_MINS = 60;
```
Comment (abhipsaMisra) [inline iothub/device/src/Authentication/Security/SharedAccessSignatureBuilder.cs:19]: The SDK should already be maintaining a default ttl reference somewhere. Reuse that instead of adding a new const again.

File: iothub/device/src/Authentication/Security/SharedAccessSignatureBuilder.cs [hunk 6]
```diff
@@ -145,5 +146,38 @@ internal static string Sign(string requestString, string key)
...
+
+        internal static string GetDeviceToken(string hostname, string deviceId, string key, long expiry = 0)
+        {
+            return GetToken(GetDeviceResourceURI(hostname, deviceId), key, expiry);
+        }
+
+        internal static string GetDeviceToken(string hostname, string deviceId, string moduleId, string key, long expiry = 0)
+        {
+            return GetToken(GetDeviceResourceURI(hostname, deviceId, moduleId), key, expiry);
+        }
```
Comment (abhipsaMisra) [inline iothub/device/src/Authentication/Security/SharedAccessSignatureBuilder.cs:171]: Instead of creating additional overloads for each of these methods you can use default values for params and have a single implementation.

File: iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs (hunk 4 above)
Comment (rido-min) [reply iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs:56]: afaik `OnBehalfOf` only works for `$edgehub`, so I agree we can have a better class name.

How about: `ClientAuthenticationForEdgeHubOnBehalfOf` ?

File: iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs (hunk 2 above)
Comment (patilsnr) [reply iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefreshBehalfOf.cs:15]: required for using a nullable operator, it's been removed and replaced by a default value in the method call
